crates/rpc/src/versioning.rs

```diff
@@ -0,0 +1,370 @@
+
+    #[inline]
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // Do not delegate to the inner service to avoid locking
+        // This if fine because we don't use more middleware and
+        // the internal service of the `jsonrpsee` server just returns
+        // `Poll::Ready(Ok(()))`
+        Poll::Ready(Ok(()))
+    }
+
+    /// Attempts to do as little error handling as possible:
+    /// - if has to manage an error condition tries to do it consistently with the inner service,
+    /// - otherwise let the inner service do it, so that there are less cases in which we have to
+    ///   care for consistency.
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        let inner = self.inner.clone();
+        let max_request_body_size = self.max_request_body_size;
+
+        let prefixes = if req.method() == Method::POST {
+            match req.uri().path() {
+                // An empty path "" is treated the same as "/".
+                // However for a non-empty path adding a trailing slash
+                // makes it a different path from the original,
+                // that's why we have to account for those separately.
+                "/" | "/rpc/v0.2" | "/rpc/v0.2/" => {
+                    Some(&[("starknet_", "v0.2_"), ("pathfinder_", "v0.1_")][..])
+                }
+                "/rpc/v0.3" | "/rpc/v0.3/" => Some(&[("starknet_", "v0.3_")][..]),
+                "/rpc/pathfinder/v0.1" | "/rpc/pathfinder/v0.1/" => {
+                    Some(&[("pathfinder_", "v0.1_")][..])
+                }
+                _ => return Box::pin(std::future::ready(Ok(response::not_found()))),
+            }
+        } else {
+            None
+        };
+
+        match prefixes {
+            Some(prefixes) => {
+                let fut = async move {
+                    // Retain the parts to then later recreate the request
+                    let (parts, body) = req.into_parts();
+
+                    let (body, is_single) =
+                        match read_body(&parts.headers, body, max_request_body_size).await {
+                            Ok(x) => x,
+                            Err(GenericTransportError::TooLarge) => {
+                                return Ok(response::too_large(max_request_body_size))
+                            }
+                            Err(GenericTransportError::Malformed) => {
+                                return Ok(response::malformed())
+                            }
+                            Err(GenericTransportError::Inner(_)) => return Ok(response::internal()),
+                        };
+
+                    let body = if is_single {
+                        let mut request: jsonrpsee::types::Request<'_> =
+                            serde_json::from_slice(&body).unwrap();
+                        prefix_method(&mut request, prefixes);
```

Comment:
It's still a bit sad that this is the only way to implement this...

Reply:
I address this in the final comment.